### PR TITLE
push metrics to pushgateway

### LINKED ISSUE #23: Add Support for Pushgateway
It would be great if the library had support for pushing to a gateway server, like the Python client (https://github.com/prometheus/client_python#exporting-to-a-pushgateway).

### PR DESCRIPTION
**Pull request by flier**
resolve #23

### CODE REVIEW

File: lib/gateway.cc [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+    labels_ = ss.str();
+  }
+}
+
+const Gateway::labels_t Gateway::instance_label(void) {
+  char hostname[MAXHOSTNAMELEN] = {0};
+
+  if (gethostname(hostname, MAXHOSTNAMELEN - 1)) {
+    hostname[0] = 0;
```
Comment (jupp0r) [inline lib/gateway.cc]: I'd rather return an empty label set here

File: include/prometheus/gateway.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  std::future<int> AsyncPush(void) { return async_push(false); }
+
+  // PushAdd metrics to the given pushgateway.
+  int PushAdd(void) { return push(true); }
+
+  std::future<int> AsyncPushAdd(void) { return async_push(true); }
+
+  // Delete metrics from the given pushgateway.
+  int Delete(void);
```
Comment (jupp0r) [inline include/prometheus/gateway.h]: Please remove the void, it doesn't make any difference in C++ (as opposed to C).

File: lib/gateway.cc [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  if (labels) {
+    for (auto& label : *labels) {
+      ss << "/" << label.first << "/" << label.second;
+    }
+  }
+
+  collectables_.push_back(std::make_pair(collectable, ss.str()));
+}
+
+std::stringstream Gateway::job_uri(void) const {
```
Comment (jupp0r) [inline lib/gateway.cc]: why not return a string on this method?

File: lib/gateway.cc [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    futures.push_back(std::async(std::launch::async, [&] {
+      return replacing ? session.Post() : session.Put();
+    }));
+  }
+
+  return std::async(std::launch::async, [&] {
+    for (auto& future : futures) {
+      auto res = future.get();
+
+      if (res.status_code > 400) {
```
Comment (jupp0r) [inline lib/gateway.cc]: You are missing the 400 (bad request) error here, should be `>=`

File: tests/integration/sample_server.cc [hunk 5]
```diff
@@ -13,6 +14,11 @@ int main(int argc, char** argv) {
   // create an http server running on port 8080
   Exposer exposer{"127.0.0.1:8080"};
 
+  // create a push gateway
+  auto labels = Gateway::instance_label();
+
+  Gateway gateway{"127.0.0.1:9091", "sample_server", &labels};
```
Comment (jupp0r) [inline tests/integration/sample_server.cc]: this will break the integration test run in CI, unless you also deploy a push gateway for the test.

**Comment by gjasny** [comment]
Thanks for your contribution. I cherry-picked you changes on top of #123 and filed #124 for review.

File: include/prometheus/gateway.h [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+#include <cpr/cpr.h>
+
+#include "histogram.h"
+#include "registry.h"
+
+namespace prometheus {
+
+class Gateway {
+ public:
+  typedef std::map<std::string, std::string> labels_t;
```
[the remark this answers is not in the document]
Comment (gjasny) [reply include/prometheus/gateway.h]: fixed in #124

File: lib/gateway.cc [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+#include "prometheus/gateway.h"
+#include "prometheus/serializer.h"
+#include "prometheus/text_serializer.h"
+
+namespace prometheus {
+
+const char* CONTENT_TYPE = "text/plain; version=0.0.4; charset=utf-8";
+
+Gateway::Gateway(const std::string& uri, const std::string jobname,
+                 const labels_t* labels, const std::string username,
```
[the remark this answers is not in the document]
Comment (gjasny) [reply lib/gateway.cc]: fixed in #124

File: lib/gateway.cc (hunk 1 above)
Comment (gjasny) [reply lib/gateway.cc]: fixed in #124

File: include/prometheus/gateway.h (hunk 2 above)
Comment (gjasny) [reply include/prometheus/gateway.h]: fixed in #124

File: lib/gateway.cc (hunk 3 above)
Comment (gjasny) [reply lib/gateway.cc]: I now calculate the value in the ctor. See #124

File: lib/gateway.cc (hunk 4 above)
Comment (gjasny) [reply lib/gateway.cc]: fixed in #124

File: tests/integration/sample_server.cc (hunk 5 above)
Comment (gjasny) [reply tests/integration/sample_server.cc]: I split it into pull mode `sample_server`(as it was before) and push mode `sample_client`

**Pull request closed without merging**